### DISCUSSION TITLE
Coroutines in Controllers and Exception Handling

### DISCUSSION THREAD

**Discussion by srinunagulapalli** [micronaut-projects/micronaut-core#10784] [General]
In our micronaut-kotlin project, we are making the controller endpoints suspendable by using the `suspend` keyword and are executing the code inside the controller endpoint by starting a `coroutineScope` like below:
```@Get("/statusDelayed")
suspend fun statusDelayed() {
    coroutineScope{
        // code execution block
    }
}
```

We are not throwing exceptions anywhere down the line in the application and using Arrow's `Either` to wrap any errors/exceptions. When the execution reaches back the controller, what happens if we unwrap this exception from the `Either` and throw the exception in the controller?

1. How does micronaut handle under the hood when a `coroutineScope` is used in controller endpoint? will it launch a coroutine under the hood? 
2. Will the exception thrown in the controller endpoint have an impact on the other controller endpoints (that would have already been launched if micronaut starts a coroutine for each controller endpoint)? 
3. Will a GlobalExceptionHandler configured with `@Error(global = true)` handle gracefully and catch immediately if an exception is thrown from a controller endpoint?
4. What is the recommended way to handle exceptions and errors in a micronaut kotlin project that deals with coroutines.

**Comment by yawkat** [comment]
In general, suspend methods are handled exactly like reactive methods. So if a controller suspends and then throws an exception (like you seem to describe), that exception will be handled as if the controller returned a `Mono.error(exc)`. Which means it'll be handled as if the controller threw the exception directly immediately.
